src/privileged/vpnRecommender/api.js

```diff
@@ -0,0 +1,562 @@
+        if (res === "success") {
+          resolve({success: true, tiCount});
+          clearInterval(ti);
+        }
+        tiCount += 1;
+
+        if (tiCount === CP_SUCCESS_MAX_CHECK_COUNT) {
+          clearInterval(ti);
+
+          resolve({success: false, tiCount});
+        }
+
+      }, CP_SUCCESS_CHECK_INTERVAL);
+    });
+
+    return promise;
+  }
+
+  listenForAddonDisableOrUninstall(addonId) {
+    let handleDisableOrUninstall;
+    const that = this;
+
+    const listener = {
+      onUninstalling(addon) {
+        handleDisableOrUninstall(addon);
+      },
+      onDisabled(addon) {
+        handleDisableOrUninstall(addon);
+      },
+    };
+
+    handleDisableOrUninstall = (addon) => {
+      if (addon.id !== addonId) {
+        return;
+      }
+
+      AddonManager.removeAddonListener(listener);
+
+      that.cleanUp();
+      addon.uninstall();
+    };
+
+    AddonManager.addAddonListener(listener);
+  }
+
+  registerListeners() {
+    log("registering listeners");
+
+    this.registerCaptivePortalTrigger();
+    this.registerPrivacyHostnameTrigger();
+    this.registerCatchallTrigger();
+    this.registerStreamingHostnameTrigger();
+  }
+
+  registerCaptivePortalTrigger() {
+    const that = this;
+
+    const observer = () => {
+      if (that._isWaitingForConnection) return; // only one connection check at a time
+
+      that.waitForConnection().then((result) => {
+        that.sendTelemetry({
+          "message_type": "captive_portal_connection_check",
+          "success": String(result.success),
+          "time": String((result.tiCount + 1) * CP_SUCCESS_CHECK_INTERVAL),
+        });
+
+        if (result.success) that.tryShowNotification(TRIGGERS.CAPTIVE_PORTAL);
+        that._isWaitingForConnection = false;
+      });
+
+      that._isWaitingForConnection = true;
+    };
+
+    Services.obs.addObserver(observer, "captive-portal-login");
+    this.addCleanUpFunction(() => {
+      Services.obs.removeObserver(observer, "captive-portal-login");
+    });
+  }
+
+  registerHostnameTrigger(hostnameList, trigger) {
+    log("registering hostname trigger");
+
+    const that = this;
+
+    const progressListener = {
+      QueryInterface: ChromeUtils.generateQI(["nsIWebProgressListener",
+        "nsISupportsWeakReference"]),
+
+      onStateChange(webProgress, request, flag, status) {
+      },
+
+      onLocationChange(progress, request, uri) {
+        // This fires when the location bar changes; that is load event is confirmed
+        // or when the user switches tabs. If you use myListener for more than one tab/window,
+        // use progress.DOMWindow to obtain the tab/window which triggered the change.
+
+        try {
+          const hostname = uri.host;
+
+          const test_function = (ref_hostname) => {
+            const escaped_ref_hostname = ref_hostname.replace(".", "\\.").replace("-", "\\-");
+            const rx = new RegExp("^(.+\\.)*(" + escaped_ref_hostname + ")$");
```

Comment:
I think you can just use [`Services.eTLD.hasRootDomain`](https://searchfox.org/mozilla-central/rev/924e3d96d81a40d2f0eec1db5f74fc6594337128/netwerk/dns/nsIEffectiveTLDService.idl#134) here.

```
Services.eTLD.hasRootDomain("abc.www.en.softonic.com", "en.softonic.com")
-> true
Services.eTLD.hasRootDomain("abc.www.de.softonic.com", "en.softonic.com")
-> false
```